### PR TITLE
Add a rust crate

### PR DESCRIPTION
**Pull request by ThatCopy**
Add a rust crate. I've already published this on crates.io [link](https://crates.io/crates/do_username), but I can transfer ownership.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,6 @@
 # Ruby gem
 ruby/*.gem
 ruby/.rspec_status
+
+#Rust crate
```
Comment (MattIPv4) [inline .gitignore]:
```suggestion
# Rust crate
```
Spacing

File: .gitignore [hunk 2]
```diff
@@ -3,3 +3,6 @@
...
+target
```
Comment (MattIPv4) [inline .gitignore]:
```suggestion
target

```
newline eof

File: rust/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+A DigitalOcean-community-themed username generator.
+
+## Install
+
+Cargo.toml
+```toml
+do_username = "0.1.0"
+```
+## Example
```
Comment (MattIPv4) [inline rust/README.md:12]:
```suggestion

## Example
```
newline before heading

File: rust/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+		let adjectives = vec!["cute", "adorable", "lovable", "happy", "sandy", "bubbly", "friendly", "floating", "drifting"];
+		let mut size= vec!["large", "big", "small", "giant", "massive", "tiny", "little", "yuge"];
+		let mut verbs = vec!["swimming", "sleeping", "eating", "hiding"];
+		let colors = vec!["blue", "blueGreen", "darkCyan", "electricBlue", "greenBlue", "lightCyan", "lightSeaGreen", "seaGreen", "turquoise", "aqua", "aquamarine", "teal", "cyan", "gray", "darkBlue", "cerulean", "azure", "lapis", "navy"];
+	
+		let mut sea_list = sea_objects;
+		let mut descriptors = adjectives;
+		
+		sea_list.append(&mut sea_creatures.clone());
+		descriptors.append(&mut size);
```
Comment (MattIPv4) [inline rust/src/lib.rs]: As per the notes in the main README, these should be exported as constants in the module that anyone can access -- they should be frozen/immutable.

File: rust/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	}
+	fn up_format(arr: Vec<&str>) -> String {
+		let mut new_arr = vec![];
+		for elem in arr {
+			new_arr.push(to_title_case(elem))
+		}
+		new_arr.join("")
+	}
+
+	pub fn get() -> String {
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
	pub fn generate(size: Option<usize>) -> String {
	  let max_size = size.unwrap_or(30);
```
Method should be named `generate`, per main README, and should support an optional size param.

File: rust/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		let rand_color = random(colors);
+		
+		let rand_desc: &str;
+		if sea_creatures.contains(&rand_noun) {
+			rand_desc = random(descriptors_and_verbs);
+		} else {
+			rand_desc = random(descriptors);
+		}
+
+		if format!("{}{}{}", rand_desc, rand_color, rand_noun).len() <= 30 {
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
		if format!("{}{}{}", rand_desc, rand_color, rand_noun).len() <= max_size {
```
Use new size param

File: rust/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+			return up_format(vec![rand_desc, rand_color, rand_noun])
+		} else if format!("{}{}", rand_desc, rand_noun).len() <= 30 {
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
		} else if format!("{}{}", rand_desc, rand_noun).len() <= max_size {
```
Use new size param

File: rust/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+			return up_format(vec![rand_desc, rand_noun])
+		} else if format!("{}{}", rand_color, rand_noun).len() <= 30 {
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
		} else if format!("{}{}", rand_color, rand_noun).len() <= max_size {
```
Use new size param

File: rust/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+			return up_format(vec![rand_color, rand_noun])
+		} else {
+			return up_format(vec![rand_noun])
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
			return up_format(vec![rand_noun])[..max_size]
```
Needs to be sliced to the maximum size

File: rust/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+		}
+	}
+}
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
}

```
newline eof

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

**Comment by ThatCopy** [comment]
I have made the changes, only thing I could not fix is the public sea_list and descriptors. I couldn't figure it out, sorry.

**Comment by MattIPv4** [comment]
I should note I don't know rust at all -- but, if you have them all as vectors instead of arrays, can you not use something like https://stackoverflow.com/a/56490417/5577674 to build the final two as well?

**Comment by ThatCopy** [comment]
If I use `let`, works fine, if I switch to `const` it doesn't work. [playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b39c5db807bc49b0bfb5256e89af7182)

**Comment by MattIPv4** [comment]
Did some more googling, looks like lazy-static lets you achieve this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=08081d17d0cb95e59a5008d0eac95c6d

**Comment by ThatCopy** [comment]
Yup, now its updated.

File: .gitignore [hunk 11]
```diff
@@ -6,4 +6,8 @@ javascript/node_modules
...
+
 ruby/coverage
```
Comment (MattIPv4) [inline .gitignore]:
```suggestion
ruby/coverage

# Rust crate
rust/target

```
Add the rust specific items after ruby

File: rust/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+description = "DigitalOcean Username Generator"
+authors = ["ThatCopy <62877244+ThatCopy@users.noreply.github.com>"]
+edition = "2018"
+homepage = "https://github.com/MattIPv4/do_username/tree/master/rust#readme"
+repository = "https://github.com/MattIPv4/do_username/"
+readme = "README.md"
+
+[dependencies]
+rand = "0.7.3"
+lazy_static = "1.0.1"
```
Comment (MattIPv4) [inline rust/Cargo.toml]:
```suggestion
lazy_static = "1.0.1"

```
newline eof

File: rust/README.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+```
+
+## Example
+
+```rust
+use do_username::do_username;
+
+fn main() {
+    // Generate a random username
+    println!("{}", do_username::get());
```
Comment (MattIPv4) [inline rust/README.md]:
```suggestion
    println!("{}", do_username::generate());
```
Method was renamed

File: rust/src/lib.rs [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+	fn up_format(arr: Vec<&str>) -> String {
+		let mut new_arr = vec![];
+		for elem in arr {
+			new_arr.push(to_title_case(elem))
+		}
+		new_arr.join("")
+	}
+	lazy_static! {
+		pub static ref SEA_CREATURES: std::vec::Vec<&'static str> = vec!["walrus", "seal", "fish", "shark", "clam", "coral", "whale", "crab", "lobster", "starfish", "eel", "dolphin", "squid", "jellyfish", "ray", "shrimp", "mantaRay", "angler", "snorkler", "scubaDiver", "urchin", "anemone", "morel", "axolotl"];
+		pub static ref SEA_OBJECTS: std::vec::Vec<&'static str> = vec!["boatShip", "submarine", "yacht", "dinghy", "raft", "kelp", "seaweed", "anchor"];
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
		pub static ref SEA_OBJECTS: std::vec::Vec<&'static str> = vec!["boat", "ship", "submarine", "yacht", "dinghy", "raft", "kelp", "seaweed", "anchor"];
```
String typo

File: rust/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+		pub static ref ADJECTIVE_DESCRIPTORS: std::vec::Vec<&'static str> = vec!["cute", "adorable", "lovable", "happy", "sandy", "bubbly", "friendly", "floating", "drifting"];
+		pub static ref SIZE_DESCRIPTORS: std::vec::Vec<&'static str> = vec!["large", "big", "small", "giant", "massive", "tiny", "little", "yuge"];
```
Comment (MattIPv4) [inline rust/src/lib.rs]:
```suggestion
		pub static ref SIZE_DESCRIPTORS: std::vec::Vec<&'static str> = vec!["large", "big", "small", "giant", "massive", "tiny", "little"];
```
String typo

File: rust/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+		let rand_noun = random(SEA_LIST.to_vec());
+		let rand_color = random(COLORS.to_vec());
+		
+		let rand_desc: &str;
+		if SEA_CREATURES.contains(&rand_noun) {
+			rand_desc = random(descriptors_and_verbs);
+		} else {
+			rand_desc = random(DESCRIPTORS.to_vec());
+		}
+
```
Comment (MattIPv4) [inline rust/src/lib.rs:52]:
```suggestion
```
Consistent whitespace

File: rust/src/lib.rs [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+		pub static ref SEA_OBJECTS: std::vec::Vec<&'static str> = vec!["boatShip", "submarine", "yacht", "dinghy", "raft", "kelp", "seaweed", "anchor"];
+		pub static ref ADJECTIVE_DESCRIPTORS: std::vec::Vec<&'static str> = vec!["cute", "adorable", "lovable", "happy", "sandy", "bubbly", "friendly", "floating", "drifting"];
+		pub static ref SIZE_DESCRIPTORS: std::vec::Vec<&'static str> = vec!["large", "big", "small", "giant", "massive", "tiny", "little", "yuge"];
+		pub static ref CREATURE_DESCRIPTORS: std::vec::Vec<&'static str> = vec!["swimming", "sleeping", "eating", "hiding"];
+		pub static ref SEA_LIST: Vec<&'static str> = SEA_OBJECTS.iter().cloned().chain(SEA_CREATURES.iter().cloned()).collect();
+		pub static ref DESCRIPTORS: Vec<&'static str> = ADJECTIVE_DESCRIPTORS.iter().cloned().chain(SIZE_DESCRIPTORS.iter().cloned()).collect();
+		pub static ref COLORS: std::vec::Vec<&'static str> = vec!["blue", "blueGreen", "darkCyan", "electricBlue", "greenBlue", "lightCyan", "lightSeaGreen", "seaGreen", "turquoise", "aqua", "aquamarine", "teal", "cyan", "gray", "darkBlue", "cerulean", "azure", "lapis", "navy"];
+	}
+	pub fn generate(size: Option<usize>) -> String {
```
Comment (MattIPv4) [inline rust/src/lib.rs:39]:
```suggestion

	pub fn generate(size: Option<usize>) -> String {
```
Consistent whitespace

File: rust/src/lib.rs [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+		arr[r]
+	}
+	fn up_format(arr: Vec<&str>) -> String {
+		let mut new_arr = vec![];
+		for elem in arr {
+			new_arr.push(to_title_case(elem))
+		}
+		new_arr.join("")
+	}
+	lazy_static! {
```
Comment (MattIPv4) [inline rust/src/lib.rs:28]:
```suggestion

	lazy_static! {
```
Consistent whitespace

File: rust/src/lib.rs [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+			Some(f) => f.to_uppercase().chain(c).collect(),
+		}
+	}
+	
+	fn random(arr: Vec<&str>) -> &str {
+		let mut rng = rand::thread_rng();
+		let r = rng.gen_range(0, arr.len());
+		arr[r]
+	}
+	fn up_format(arr: Vec<&str>) -> String {
```
Comment (MattIPv4) [inline rust/src/lib.rs:20]:
```suggestion

	fn up_format(arr: Vec<&str>) -> String {
```
Consistent whitespace

File: rust/README.md [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+}
+
```
Comment (MattIPv4) [inline rust/README.md]:
```suggestion
```
Consistent whitespace

**Review by MattIPv4 — CHANGES_REQUESTED** [review]

**Review by MattIPv4 — CHANGES_REQUESTED** [review]
One final code change, we have a mix of `std::vec::Vec` and `Vec`. I think we're safe to just use `Vec` everywhere?

**Comment by ThatCopy** [comment]
AFAIK they are the same

**Comment by ThatCopy** [comment]
Almost forgot, do you want me to add you to the `authors` in `Cargo.toml` if yes, can you give me the email I should use?

**Comment by MattIPv4** [comment]
Yeah, please add `DigialOcean, Matt Cowley <me@mattcowley.co.uk>` to authors :)

**Comment by MattIPv4** [comment]
👋 One final thing, could you add a quick file on how to publish a version for posterity (see https://github.com/MattIPv4/do_username/pull/24)?

**Review by MattIPv4 — APPROVED** [review]
Lgtm!

**Comment by MattIPv4** [comment]
Tyvm for porting this over to Rust -- final thing before merging, how do we transfer ownership on crates?

(Once merged I'll do a 1.0.0 release as the changes in this PR involve a public method change from the original crate)

**Comment by ThatCopy** [comment]
I'm pretty new to this but I think I can add you by a username. You will need to make a crates.io account. 
image

**Comment by ThatCopy** [comment]
You will need to make a crates.io account. Than I can add you by your username 
image

**Comment by MattIPv4** [comment]
👍 I'm `MattIPv4` on crates

**Comment by MattIPv4** [comment]
Oki, looks like we're good to go. Will merge and attempt the 1.0.0 release. Ty again!

**Pull request merged**